REQUIREMENTS.md

```diff
@@ -0,0 +1,155 @@
+* Patch: A piece of an IPLD DAG. Blocks contain patches.
+* Node: A node is a *point* in an IPLD DAG (object, array, number, etc.).
+* Link: A link is an IPLD Node that points to another IPLD Node.
+* Path: A paths a human readable pointer to an IPLD Node.
+
+## Linked
+
+IPLD must support linking to any IPLD node (even if the node is in the middle of
+a block). That is, IPLD must support arbitrary IPLD paths in links.
+
+**Motivation:** Considering this in the context of programming languages, not
+being able to *store* a pointer to a struct *inside* of another struct would be
+severely limiting.
+
+NOTE: We don't currently support arbitrary paths but, in the context of
+programming, we really need to.
+
+## Immutable
+
+IPLD links must be immutable. We'll likely define a mutable link spec on top of
+IPLD but there needs to be an immutable layer at the bottom.
+
+**Motivation:** *Having* an immutable layer is important for a lot of analysis,
+memoization, type checking, etc.
+
+## No Non-Local Reasoning
+
+An IPLD block should never be interpreted in the context of *anything* not
+contained in the block (and CID).
+
+For example, assuming we add support for relative links, the following
+definition of `foo` would not be a valid IPLD block:
+
+```
+var foo = {
+  "baz": {"/": "../../baz"}
+}
+var bar = {
+  "foo": CidOf(foo),
+  "baz": "something"
+}
+
+// resolution throug block `foo` depends on block `bar`.
+Resolve("/ipld/${CidOf(bar)}/foo/baz/")
+```
+
+For the same reason, IPLD links can't rely on an authority (e.g., a blockchain).
+
+**Motivation:** IPLD needs to be easy to reason about.
+
+**Caveat:**
+
+We *may* want to relax this if we want to move schemas into separate,
+deduplicated blocks (referenced by CID). If we do that, we'd need to fetch a
+block's schema before being able to interpret the it.
+
+However, we need to *thoroughly* discuss any changes to this requirement.
+
+1. The space savings may not be worth it given the size of CIDs (>40 bytes),
+   compression, smart transports, and smart datastores.
+2. This change would introduce some weird interface complexities and potential
+   network dependencies.
+
+## No Cycles
+
+IPLD links must not be cyclic, even if we add support for relative links.
+
+**Motivation:**
+
+1. Security: Forbidding link cycles ensures that any graph traversal terminates.
+   This makes it easier to correctly and securely implement some graph
+   algorithms, even on potentially untrusted data in a distributed system.
+2. Consistency: Without complex hacks, it's impossible to create link cycles
+   *between* hash-linked blocks. Allowing link cycles one one level (e.g., in
+   relative links within a single block) but not on a larger scale (between
+   blocks) is inconsistent (even though there's nothing we can do about it).
+
+## Stable Pathing
+
+An IPLD path always means the same thing, everywhere, every time. Importantly,
```

Comment:
We may want to qualify this to a particular version of "Path." We're talking about a few iterations on pathing that *will* change how it works.